adam/visualization/positioning.py

```diff
@@ -0,0 +1,426 @@
+        )
+        return distance_penalty + collision_penalty
+
+    def dump_object_positions(self) -> None:
+        for (adam_object, bounding_box) in self.adam_object_to_bounding_box.items():
+            print(f"{adam_object.name} = {bounding_box.center.data}")
+
+
+class DistanceFromOriginPenalty(nn.Module):
+    def __init__(self) -> None:  # pylint: disable=useless-super-delegation
+        super().__init__()
+
+    def forward(  # pylint: disable=arguments-differ
+        self, bounding_box: AxisAlignedBoundingBox
+    ) -> torch.Tensor:
+        return bounding_box.center_distance_from_point(ORIGIN)
+
+
+class CollisionPenalty(nn.Module):
+    def __init__(self):  # pylint: disable=useless-super-delegation
+        super().__init__()
+
+    def forward(  # pylint: disable=arguments-differ
+        self,
+        bounding_box_1: AxisAlignedBoundingBox,
+        bounding_box_2: AxisAlignedBoundingBox,
+    ) -> torch.Tensor:
+
+        # get face norms from one of the boxes:
+        face_norms = bounding_box_2.get_face_norms()
+
+        return CollisionPenalty.overlap_penalty(
+            CollisionPenalty.get_min_max_overlaps(
+                CollisionPenalty.get_min_max_corner_projections(
+                    bounding_box_1.corner_onto_axes_projections(face_norms)
+                ),
+                CollisionPenalty.get_min_max_corner_projections(
+                    bounding_box_2.corner_onto_axes_projections(face_norms)
+                ),
+            )
+        )
+
+    @staticmethod
+    def get_min_max_corner_projections(projections: torch.Tensor):
+        """
+        Retrieve the minimum and maximum corner projection (min/max extent in that dimension) for each axis
+        Args:
+            projections: Tensor(3, 8) -> corner projections onto each of three dimensions
+
+        Returns:
+            Tensor(3, 2) -> (min, max) values for each of three dimensions
+
+        """
+        check_arg(projections.shape == (3, 8))
+
+        min_indices = torch.min(projections, 1)
+        max_indices = torch.max(projections, 1)
+        # these are tuples of (values, indices), both of which are tensors
+
+        # helper variable for representing dimension numbers
+        # see https://github.com/pytorch/pytorch/issues/24807 re: pylint issue
+        dims = torch.tensor([0, 1, 2], dtype=torch.int)  # pylint: disable=not-callable
+        # select the indexed items (from a 24 element tensor)
+        minima = torch.take(projections, min_indices[1] + (dims * 8))
+        maxima = torch.take(projections, max_indices[1] + (dims * 8))
+        # stack the minim
+        return torch.stack((minima, maxima), 1)
+
+    @staticmethod
+    def get_min_max_overlaps(
+        min_max_proj_0: torch.Tensor, min_max_proj_1: torch.Tensor
+    ) -> torch.Tensor:
+        """
+        Given min/max corner projections onto 3 axes from two different objects,
+        return an interval for each dimension representing the degree of overlap or
```

Comment:
Comment on how the returned values should be interpreted in terms of overlap/separation